contrib/mod_tls.c

```diff
@@ -13005,6 +13024,78 @@ MODRET tls_any(cmd_rec *cmd) {
   return PR_DECLINED(cmd);
 }
 
+MODRET tls_pre_xfer(cmd_rec *cmd) {
+#if defined(TLS1_3_VERSION)
+  SSL_SESSION *ssl_session;
+  long sess_created, sess_expires, sess_remaining = 0;
+  int request_new_ticket = FALSE, ssl_version;
+  time_t now;
+#endif /* TLS1_3_VERSION */
+
+  if (tls_engine == FALSE) {
+    return PR_DECLINED(cmd);
+  }
+
+  if (!(tls_flags & TLS_SESS_ON_CTRL)) {
+    return PR_DECLINED(cmd);
+  }
+
+  if (ctrl_ssl == NULL) {
+    return PR_DECLINED(cmd);
+  }
+
+#if defined(TLS1_3_VERSION)
+  ssl_session = SSL_get_session(ctrl_ssl);
+
+  /* Ideally I would use SSL_SESSION_has_ticket, available since OpenSSL 1.1.x,
+   * rather than simply gating on the session protocol version, but that does
+   * not work as expected in local regression testing, whereas using the
+   * protocol version does work.
+   */
+  ssl_version = SSL_SESSION_get_protocol_version(ssl_session);
+  if (ssl_version != TLS1_3_VERSION) {
+    return PR_DECLINED(cmd);
+  }
+
+  sess_created = SSL_SESSION_get_time(ssl_session);
+  sess_expires = SSL_SESSION_get_timeout(ssl_session);
+  now = time(NULL);
+
+  if ((sess_created + sess_expires) >= now) {
+    sess_remaining = (unsigned long) ((sess_created + sess_expires) - now);
+  }
+
+  pr_trace_msg(trace_channel, 21,
+    "%s: control channel TLSv1.3 session lifetime: %ld %s",
+    (const char *) cmd->argv[0], sess_remaining,
+    sess_remaining != 1 ? "seconds" : "second");
+
+  /* If the control session expires in 10 seconds or less, request a new one
+   * before proceeding with the data transfer (Issue #1931).
+   */
+  if (sess_remaining <= 10) {
+    request_new_ticket = TRUE;
+  }
+
+# if defined(OPENSSL_VERSION_MAJOR) && \
+     OPENSSL_VERSION_MAJOR >= 3
+  if (request_new_ticket == TRUE) {
+    if (SSL_new_session_ticket(ctrl_ssl) == 1) {
```

Comment:
I will experiment with using [`SSL_key_update`](https://docs.openssl.org/master/man3/SSL_key_update/) instead here, as is done [elsewhere](https://github.com/proftpd/proftpd/blob/master/contrib/mod_tls.c#L4140), to see if that also fixes things.   Why?  `SSL_key_update` first appeared in OpenSSL 1.1.x, meaning that this fix would work for older OpenSSL installations as well.